technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+#### Event: K8ssandraCluster created
+
+##### k8ssandra controller
+
+* Start reconciliation loop when K8ssandraCluster object is created
+* Creates a CassandraCluster
+  * Update status
+* Creates a Reaper
+  * Update status
+* Creates a Stargate
+  * Update status
+* Creates a MonitoringConfiguration
+  * Update status
+* End reconciliation loop
+
+#### Event: CassandraCluster created
+
+##### cassandra cluster controller
+
+* Start reconciliation loop when CassandraCluster object is created
+* Create CassandraDatacenter
+  * Update status
+* End reconciliation loop
+
+#### Event: Reaper created
+
+##### reaper controller
+
+* Start reconciliation loop when Reaper object is created
+* Creates reaper Service
+* Requeue request until CassandraDatacenter is ready
+
+#### Event: Stargate created
+
+##### stargate controller
+
+* Start reconciliation loop when Stargate object is created
+* Creates stargate Service
+* Creates ServiceMonitor
+* Creates a GrafanaDashboard
+* Requeue request until CassandraDatacenter is ready
+
+#### Event: MonitoringConfiguration created
+
+##### monitoring controller
+
+* Start reconciliation loop when MonitoringConfiguration created
+* Create Grafana object
+* Create GrafanaDataSource object
+* Create Prometheus object
+* End reconciliation loop
+
+#### Event: CassandraDatacenter ready
+
+##### cassandra cluster controller
+
+* Update status to reflect state of CassandraCluster
+* End reconciliation loop
+
+##### reaper controller
+
+* Start reconciliation loop after requeue
+* Create Job to initialize reaper keyspace
+* Requeue request until job finishes
+
+##### stargate controller
+
+* Start reconciliation loop after requeue
+* Create stargate Deployment
+* Requeue request until Deployment ready
+
+#### Event: CassandraCluster status updated
+
+##### k8ssandra controller
+
+* Start reconciliation loop
+  * Update status to reflect state of CassandraCluster
+* End reconciliation loop
+
+#### Event: Reaper schema job completed
+
+##### reaper controller
+
+* Create Deployment
+* Requeue request until Deployment ready
+* End reconciliation loop
+
+#### Event: Stargate ready
+
+##### k8ssandra controller
+
+* Update status to reflect state of Stargate
+
+#### Event: Reaper ready
+
+##### k8ssandra controller
+
+* Update status to reflect state of Reaper
+
+#### Event: Grafana and Prometheus ready
+
+##### monitoring controller
+
+* Update status of MonitoringConfiguration to reflect state of Grafana and Prometheus
+
+#### Event: MonitoringConfiguration status updated
+
+##### k8ssandra controller
+
+* Update status to reflect state of MonitoringConfiguration
+
+#### Notes
+
+The k8ssandra controller creates several objects:
+
+* CassandraCluster
+* Reaper
+* Stargate
+* MonitoringConfiguration
+
+Whenever any of these objects is updated, it should trigger a reconciliation of the k8ssandra controller. In the reconciliation loop, the status of each object will essentially be pushed up into the K8ssandraCluster object.
```

Comment:
With the push up of the status you mention, we would then be able to inspect just the top-level object to get status information for the whole deployment from the top-down?

Reply:
Yup, that's the idea. Then you can inspect child objects for additional details. I know that I said multi-cluster will be covered in a separate, follow-up doc, but I want to point out that this would apply in a multi-cluster deployment as well. That top-level status would provide information across all k8s clusters.

Reply:
I like that idea, a rolled-up status that includes all k8s clusters would be awesome.

Reply:
Is this push two-way? So when K8ssandraCluster object is modified, those changes too are pushed down?